provider/mesh/peer.go

```diff
@@ -12,35 +13,107 @@ import (
 	"github.com/weaveworks/mesh"
 )
 
+type replaceFile struct {
+	*os.File
+	filename string
+}
+
+func (f *replaceFile) Close() error {
+	if err := f.File.Sync(); err != nil {
+		return err
+	}
+	if err := f.File.Close(); err != nil {
+		return err
+	}
+	return os.Rename(f.File.Name(), f.filename)
+}
+
+func openReplace(filename string) (*replaceFile, error) {
+	tmpFilename := fmt.Sprintf("%s.%s", filename, utcNow().Format(time.RFC3339Nano))
+
+	f, err := os.Create(tmpFilename)
+	if err != nil {
+		return nil, err
+	}
+
+	rf := &replaceFile{
+		File:     f,
+		filename: filename,
+	}
+	return rf, nil
+}
+
 // NotificationInfos provides gossiped information about which
 // receivers have been notified about which alerts.
 type NotificationInfos struct {
-	st     *notificationState
-	send   mesh.Gossip
-	logger log.Logger
+	st        *notificationState
+	send      mesh.Gossip
+	retention time.Duration
+	snapfile  string
+	logger    log.Logger
+	stopc     chan struct{}
 }
 
-func NewNotificationInfos(logger log.Logger) *NotificationInfos {
-	return &NotificationInfos{
-		logger: logger,
-		st:     newNotificationState(),
+func NewNotificationInfos(logger log.Logger, retention time.Duration, snapfile string) (*NotificationInfos, error) {
+	ni := &NotificationInfos{
+		logger:    logger,
+		stopc:     make(chan struct{}),
+		st:        newNotificationState(),
+		retention: retention,
+		snapfile:  snapfile,
+	}
+	f, err := os.Open(snapfile)
+	if os.IsNotExist(err) {
+		return ni, nil
+	}
+	if err != nil {
+		return nil, err
 	}
+	defer f.Close()
+
+	return ni, ni.st.loadSnapshot(f)
 }
 
 // Register the given gossip channel.
 func (ni *NotificationInfos) Register(g mesh.Gossip) {
 	ni.send = g
 }
 
+const maintenanceInterval = 15 * time.Minute
```

Comment:
Should be a flag eventually. (Add a TODO if you don't want to add a flag right now.)